### ISSUE TITLE
Refactoring the code

### ISSUE THREAD

**Issue by hadi-gharibi** [maciejkula/spotlight#163]
Hello,
I was wondering if you have any plan to refactor the code and make it more look like a framework like fastai, ignite, etc. 
Right now we have to implement the training loop for each new model and this is not efficient. If we could use callbacks, it would be more easy to use more complicated training approaches(like anneal learning rate) and at the same time use something like TensorFX and even logging without headache.
